src/user.rs

```diff
@@ -424,25 +408,165 @@ async fn on_room_message(
 
 /// This function returns homeserver domain and url, ex:
 ///  - get_homeserver_url("matrix.domain.com") => ("matrix.domain.com", "https://matrix.domain.com")
-fn get_homeserver_url<'a>(homeserver: &'a str, protocol: Option<&'a str>) -> (&'a str, String) {
+fn get_homeserver_url(homeserver: &str, protocol: Option<&str>) -> (String, String) {
     let regex = Regex::new(r"https?://").unwrap();
     if regex.is_match(homeserver) {
         let parts: Vec<&str> = regex.splitn(homeserver, 2).collect();
-        (parts[1], homeserver.to_string())
+        (parts[1].to_string(), homeserver.to_string())
     } else {
         let protocol = protocol.unwrap_or("https");
-        (homeserver, format!("{protocol}://{homeserver}"))
+        (homeserver.to_string(), format!("{protocol}://{homeserver}"))
+    }
+}
+
+async fn get_client(homeserver_url: &str, retry_enabled: bool) -> Option<Client> {
+    let instant = Instant::now();
+
+    let (_, homeserver) = get_homeserver_url(homeserver_url, None);
+
+    log::info!("Attempt to create a client");
+
+    let timeout = Duration::from_secs(30);
+
+    let request_config = if retry_enabled {
+        RequestConfig::new().retry_timeout(timeout)
+    } else {
+        RequestConfig::new().disable_retry().timeout(timeout)
+    };
+
+    let client = Client::builder()
+        .request_config(request_config)
+        .homeserver_url(homeserver)
+        .build()
+        .await;
+
+    match client {
+        Ok(client_value) => {
+            log::info!("New client created {}", instant.elapsed().as_millis());
+
+            Some(client_value)
+        }
+        Err(_) => {
+            println!("got request config error {}", client.err().unwrap());
+            log::info!("Failed to create client");
+            None
+        }
```

Comment:
Before I meant to use the variable `client` here and change the function's signature to `Result<Client, ...>` instead. This is not critical, but a suggestion.

Reply:
I think I prefer the current solution for logging purposes mostly